### DISCUSSION TITLE
Custom queries?

### DISCUSSION THREAD

**Discussion by keithf4** [pgexporter/pgexporter#12] [Ideas]
Any plans to include options to provide custom queries/metrics?

I've found with playing with different exporters that providing the queries built into the source always creates a problem at some point with a new PG release. Catalogs change, metrics then break and you have to wait for upstream to provide a fix. Even within a point release, sometimes catalog permissions could change. 

Which metrics to provide is always a very opinionated topic, so combined with above issue, I always thought a better model for any metric collection would be to have the metrics/queries be separate from the start and the default config just require that you set it up with some good default metrics that upstream can certainly provide. But always leave them optional.

**Comment by jesperpedersen** [comment]
Hi Keith.

Yes, there are definitely plans to include support for custom queries. Like you said, some people are looking for specific queries that wouldn't be covered by a default monitoring install.

However, I want to get the type, help and version infrastructure and additional core metrics in place first before I'll look at this area.

Feel free to share your ideas and implementation on how you best see this area integrated.

Thanks !

**Comment by keithf4** [comment]
Big picture idea at the moment would be to just provide the default queries/metrics as an additional file that is fed in via a `--query_file` or `--metrics` sort of option. Or if you do want to have those be built in by default, at least make them optional via a `--disable-builtin-metrics` flag. Would allow people to keep working with the tool despite any upstream changes from core postgres, especially if they just want to use their own.

Allow one or more custom query files to be provided to the exporter. One of the important options for a metric would be which database(s) to run a given query on. Would make it easy to support both global metrics that are the same on any database (so it would only need to run against one) and specific metrics that are only available on a per-database basis (table stats, etc) from a single exporter instance.

**Reply by jesperpedersen** [reply]
An idea could be to have one or more metrics defined inside the "file", and have pgexporter pick up a directory with these files. The "file" would define the query, type, help, version and database(s) to run it on.

"Native" queries could be disabled by a `core = off` and `extended = off` flag in the configuration.

I'm not sold on YAML as a format, but it could be an option...

**Reply by bassamadnan** [reply]
@jesperpedersen have we ever considered using `.sql` files directly for these metrics? The idea just striked me, i do understand that we would've chosen YAML for some extra fields to avoid comparision logics and handling variables.

**Reply by jesperpedersen** [reply]
@bassamadnan We have to be a bit careful due to Little Bobby Tables (https://xkcd.com/327/)

**Reply by bassamadnan** [reply]
@jesperpedersen a pg monitor role would be centered around reads only right?  are we being careful of cases where a user may not be running pgexporter on more elivated roles?

**Reply by jesperpedersen** [reply]
@bassamadnan Yes, we should assume a `pg_monitor` role, and the YAML/JSON files are ok... but ...
@Userfrom1995 Think about how you can help @bassamadnan with SQL injection problems...
